crates/bevy_ecs/src/change_detection.rs

```diff
@@ -65,6 +72,25 @@ pub trait DetectChanges {
     /// [`SystemChangeTick`](crate::system::SystemChangeTick)
     /// [`SystemParam`](crate::system::SystemParam).
     fn last_changed(&self) -> u32;
+
+    /// Sets `self` to `value`, if and only if `*self != *value`
+    ///
+    /// `T` is the type stored within the smart pointer (e.g. [`Mut`] or [`ResMut`]).
+    ///
+    /// This is useful to avoid accidentally triggering change detection when no change is made,
+    /// as changes are usually deemed to be made when [`DerefMut`] is used.
+    #[inline]
+    fn set_if_differs<T>(&mut self, value: T)
+    where
+        Self: Deref<Target = T> + DerefMut<Target = T>,
+        T: PartialEq,
+    {
+        let immutable_ref: &T = self.deref();
+        if *immutable_ref != value {
```

Comment:
```suggestion
        let shared_ref: &T = self.deref();
        if *shared_ref != value {
```

We shouldn't use the naïve names here.

It's also not clear to me why we're being so explicit here with the variables either? Especially not within the `if` branch - I can see that for this one you want to make it obvious that change detection isn't triggered.